spec/ics-002-client-semantics/README.md

```diff
@@ -473,27 +597,97 @@ function checkValidityAndUpdateState(
     clientState.verifiedRoots[sequence] = header.commitmentRoot
 }
 
-// state membership verification function defined by the client type
-function verifyMembership(
+function verifyClientConsensusState(
   clientState: ClientState,
-  sequence: uint64,
+  height: uint64,
+  prefix: CommitmentPrefix,
+  proof: CommitmentProof,
+  clientIdentifier: Identifier,
+  consensusState: ConsensusState) {
+    path = applyPrefix(prefix, "clients/{clientIdentifier}/consensusState")
+    abortTransactionUnless(!clientState.frozen)
+    return clientState.verifiedRoots[sequence].verifyMembership(path, consensusState, proof)
+}
+
+function verifyConnectionState(
+  clientState: ClientState,
+  height: uint64,
+  prefix: CommitmentPrefix,
   proof: CommitmentProof,
-  path: Path,
-  value: Value) {
+  connectionIdentifier: Identifier,
+  connectionEnd: ConnectionEnd) {
+    path = applyPrefix(prefix, "connection/{connectionIdentifier}")
     abortTransactionUnless(!clientState.frozen)
-    return clientState.verifiedRoots[sequence].verifyMembership(path, value, proof)
+    return clientState.verifiedRoots[sequence].verifyMembership(path, connectionEnd, proof)
 }
 
-// state non-membership function defined by the client type
-function verifyNonMembership(
+function verifyChannelState(
   clientState: ClientState,
+  height: uint64,
+  prefix: CommitmentPrefix,
+  proof: CommitmentProof,
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  channelEnd: ChannelEnd) {
+    path = applyPrefix(prefix, "ports/{portIdentifier}/channels/{channelIdentifier}")
```

Comment:
question: why do we want to expose the paths hardcoded on the client?

Reply:
These are example implementations, not necessarily required.

They will vary per client type. See https://github.com/cosmos/ics/issues/286